server/pulp/plugins/rsync/publish.py

```diff
@@ -0,0 +1,522 @@
+        return cmd
+
+    def make_authentication(self):
+        """
+        Returns a string representation of command for authenticating against a remote server.
+
+        :return: str. e.g., '-e ssh -l ssh_login_user -i /key_path'
+        :rtype: str
+        """
+        if self.get_config().flatten()["remote"]['auth_type'] == 'local':
+            return ()
+        ssh_parts = []
+        for arg in self.make_ssh_cmd():
+            if " " in arg:
+                ssh_parts.append('"%s"' % arg)
+            else:
+                ssh_parts.append(arg)
+
+        return ['-e', " ".join(ssh_parts)]
+
+    def make_full_path(self, relative_path):
+        """
+        Returns full path of a relative path on the remote server.
+
+        :param relative_path: relative path
+        :type relative_path: str
+        """
+        return os.path.join(self.get_config().flatten()["remote"]['root'], relative_path)
+
+    def make_destination(self, relative_path):
+        """
+        Parse from self.config information to make up a hostname and remote path used for rsync
+        command.
+
+        :return: str of the combination of user, host, and dir.
+        :rtype:  str
+        """
+        if self.get_config().flatten()["remote"]['auth_type'] == 'local':
+            return self.get_config().flatten()["remote"]['root']
+        relative_path = relative_path.lstrip("/")
+
+        user = self.get_config().flatten()["remote"]['ssh_login']
+        host = self.get_config().flatten()["remote"]['host']
+        remote_root = self.get_config().flatten()["remote"]['root']
+        return '%s@%s:%s' % (user, host, os.path.join(remote_root, relative_path))
+
+    def call(self, args, include_args_in_output=True):
+        """
+        A wrapper around kobo.shortcuts.run. If ssh_exchange_identification or
+        max-concurrent-connections exceptions are thrown by ssh, up to 10 retries follows.
+
+        :return: (boolean indicating success or failure and std_out + std_err)
+        :rtype: tuple of boolean and string
+        """
+        for t in xrange(10):
+            rv, out = kobo.shortcuts.run(cmd=args, can_fail=True)
+            possible_known_exceptions = \
+                ("ssh_exchange_identification:" in out) or ("max-concurrent-connections=25" in out)
+            if not (rv and possible_known_exceptions):
+                break
+            _logger.info("Connections limit reached, trying once again in thirty seconds.")
+            time.sleep(30)
+        if include_args_in_output:
+            message = "%s\n%s" % (args, out)
+        else:
+            message = out
+        return (rv == 0, message)
+
+    def make_rsync_args(self, files_from, source_prefix, dest_prefix, exclude=None):
+        """
+        Creates a list of arguments for the rsync command
+
+        :param files_from: list of files to rsync
+        :type files_from: list of strings
+        :param source_prefix: location of files to rsync
```

Comment:
I assume since "prefix" is in the name that this location is relative to something else? Same question for ``dest_prefix``.